### ISSUE TITLE
Tag template of arrow function infers wrong type

### ISSUE THREAD

**Issue by danieldietrich** [microsoft/TypeScript#36700]
**TypeScript Version:**  3.8.0-dev.20200208

**Search Terms:** tag function, tagged template, template string

**Code**

```ts
{
    function tag(strings: TemplateStringsArray, ...args: unknown[]): [TemplateStringsArray, unknown[]] {
        return [strings, args];
    }

    // ✅ inferred type: [TemplateStringsArray, unknown[]]
    const t = tag`123`;

    // console.log(t[2]); // it is correct that this would not compile
}

{
    const tag = (strings: TemplateStringsArray, ...args: unknown[]) => [strings, args];

    // 🛑 inferred type: (TemplateStringsArray | unknown[])[]
    const t = tag`123`;

    console.log(t[2]); // should not compile, but it does!
}
```

**Expected behavior:**

The `console.log(t[2])` statement should produce the following error:

```
Tuple type '[TemplateStringsArray, unknown[]]' of length '2' has no element at index '2'. ts(2493)
```

**Actual behavior:**

The example does compile (but it shouldn't).

**Playground Link:** [here](https://www.typescriptlang.org/play/?ssl=9&ssc=5&pln=9&pc=71#code/N4KABBYGYK4HYGMAuBLA9nMSCGBzAFAM5IBOKcuhAXGACoCmAtgA4A22S9AyqeZQIIkS2AJ4AaMADpp2EpRrwA1nDQB3OAG0AugEoaGhi3aceZCoUHDxYJSvXatYUJBdgS9JDBKYNxM5QlZSi0AbnBIAF8QcIgAeliwQFByMHIoeiF6ABMsEWZ6fUM2Dm5ec0tRCVs1TS0tGLAEDGIsMABeLDwAAwBGACYAZk6w+viGprRWeklWNAIkDV7dELBRlCQUwjGM5CwACw49lE3VNBhWbJV1xpYUSZAokGdIRrhmnFw2sCJS+TomIpMPwsQgqUhkcmoNjgymq2h0bQAfGBfEDAhDQtEXKNALwbgEW9lJwNIZbJIXL5L6FYwlfzAqxgAA+UJh9l02nqLzen3ePQGQ0xz3Gk2ms3w80WOmWo0Iu1O5zAlzGN0mEgARjB1mswJk0PRCABCe5AA)

**Related Issues:** []

**Comment by Zemnmez** [comment]
It looks like this type inferrence is at least consistent with calling the same function _without_ using template syntax:

```
{
    function tag(strings: TemplateStringsArray, ...args: unknown[]): [TemplateStringsArray, unknown[]] {
        return [strings, args];
    }

    // ✅ inferred type: [TemplateStringsArray, unknown[]]
    const t = tag`123`;

    // console.log(t[2]); // it is correct that this would not compile
}

{
    const tag = (strings: TemplateStringsArray, ...args: unknown[]) => [strings, args];

    // 🛑 inferred type: (TemplateStringsArray | unknown[])[]
    const t = tag`123`;

    console.log(t[2]); // should not compile, but it does!
}

{
    const tag = (strings: TemplateStringsArray, ...args: unknown[]) => [strings, args];

    // 🛑 inferred type: (TemplateStringsArray | unknown[])[]
    const t = tag(Object.assign(["123"], {raw: ["123"]}));

    console.log(t[2]); // should not compile, but it does!
}
```

[Playground Link](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=29&pc=1#code/N4WAUABFEGYK4DsDGAXAlgewRFBDA5gBQDOKATmgvsQFwQAqApgLYAOANriowMrmXUAgmTK4AngBoIAOlm4y1OogDWCDAHcEAbQC6ASjpambTtz4UqxYaMkQVazbp0RQkaO7KMUcMti2kLail5ah0AbnB3AF9wSOgAeniIQFByCEoYRhFGABMcMVZGQ2MOLl5+S2txKXsNbR0dOKgkLFIcCABeHAIAAwBGACYAZm6IsEaIRIhmhGIMdkZpdgwiFC1+-TCJpLQUNOIpjCzUHAALLlO0ffUMOHZctV3mtjR58BixsFd3ada8fA6ICRyooGCwSmZgVYRFUZHIFLQ7AhVLVdHoOgA+CD+SHBeHhWJuKCTQC8G4BFvbSCAyWVyKHyhUBxVMZUCUJsEAAPojkY59Lpxj9drtOn8+kMRgTvi05gslis1hsthBiCcbncIA8Ds95lIAEZwXY7CDZDCMYgAQjeBK+0AFXX+nSBLLojNK5gErJhsmkIQRNR5aPamOxLNxoVG4xJ5PSmU8NLpdEILohLMqYg5XIcdT0fMJBxmgoBf0IAHkdQArRiob3EYhofAIQhaABEA0GTZ0UmAonUhhbQ3bUT0enDuZ+0sWy0Iq3Ww8VytV9wwjwwWsYuv1aV2xtNFrA7yAA)

**Comment by danieldietrich** [comment]
That's right.

An additional type witness does solve the problem. But it should work without it.

```ts
{
    const tag = (strings: TemplateStringsArray, ...args: unknown[]) => [strings, args];

    // 🛑 inferred type: (TemplateStringsArray | unknown[])[]
    const t = tag`123`;

    console.log(t[2]); // should not compile, but it does!
}

{
    const tag = (strings: TemplateStringsArray, ...args: unknown[]): [TemplateStringsArray, unknown[]] => [strings, args];

    // 🛑 inferred type: (TemplateStringsArray | unknown[])[]
    const t = tag`123`;

    console.log(t[2]); // it is correct that this does not compile
}
```

[Playground Link](https://www.typescriptlang.org/play/?ssl=25&ssc=5&pln=25&pc=59#code/N4WAUABFEGYK4DsDGAXAlgewRFBDA5gBQDOKATmgvsQFwQAqApgLYAOANriowMrmXUAgmTK4AngBoIAOlm4y1OogDWCDAHcEAbQC6ASjpambTtz4UqxYaMkQVazbp0RQkaO7KMUcMti2kLail5ah0AbnB3AF9wSOgAeniIQFByCEoYRhFGABMcMVZGQ2MOLl5+S2txKXsNbR0dOKgkLFIcCABeHAIAAwBGACYAZm6IsEaIRIhmhGIMdkZpdgwiFC1+-TCJpLQUNOIpjCzUHAALLlO0ffUMOHZctV3mtjR58BixsFd3ada8fA6ICRyooGCwSmZgVYRFUZHIFLQ7AhVLVdHoOgA+CD+SHBeHhWJuKCTQC8G4BFvbSCAyWVyKHyhUBxVMZUCUJsEAAPojkY59Lpxj9drtOn8+kMRgTvi05gslis1hsthBiCcbncIA8Ds95lIAEZwXY7CDZDCMYgAQjeBK+0AFXX+nSBLLojNK5gErJhsmkIQRNR5BixLohLMqtj9dWc7Ux2JZuNCo3GJPJ6UynhpdLohCDzPdoY5XIcdT0fMJBxmgoBIoGwwTpZ+0sWy0Iq3Wek2k0NlwOR0FZ17XeNpvVGEeGC1jEtYCAA)

**Comment by Zemnmez** [comment]
I think this is down to typescript not detecting tuple return types at all, and therefore not being able to correctly intuit the response here. I think this is a feature request, rather than a bug then.

Consider this minimal example:

```
{
    // f: () => number[], rather than f: () => [1,2]
    const f = () => {
        return [1,2]
    }
}
```

[Playground Link](https://www.typescriptlang.org/play/#code/N4WAUABFEPQxAzAXBAFASggXgHwQHYCuAtgEYCmATgNoC6ANBJQIYAuAFlRB8-oihmx5qARnoAmWuGgQAxgHt8AZ1aJsaTLgihIM6JXKtClPqIlTdUAL7grQA)

**Comment by danieldietrich** [comment]
You are absolutely right. Without an additional type hint, type inference for a plain function behaves exactly the same. I will close this issue (and I will not create a feature request).

Many thanks for your investigation!

```
{
    // f: () => number[], rather than f: () => [1,2]
    const f = () => {
        return [1,2]
    }
}

{
    // f(): number[], perfectly consistent
    function f() {
        return [1,2]
    }
}
```

[Playground Link](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=13&pc=2#code/N4WAUABFEPQxAzAXBAFASggXgHwQHYCuAtgEYCmATgNoC6ANBJQIYAuAFlRB8-oihmx5qARnoAmWuGgQAxgHt8AZ1aJsaTLgihIM6JXKtClPqIlTdUAL7gbYcDplxEGFETJU6jAA5UE5WVYAGwBPOUUlAEsVcnxWaWgEQnxAyMUXTEc9KAMjEwgzSQTrWyA)

**Comment by Zemnmez** [comment]
np! I would love to see this inference too :)

**Comment by danieldietrich** [comment]
I think it would be risky, the change might not be source compatible regarding existing code bases. But currently I can't find a scenario where it would break existing code.

```ts
function f(ns: number[]) { }

const arr: [1, 2] = [1, 2]

f(arr); // works!
```

[Playground Link](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=5&pc=18#code/GYVwdgxgLglg9mABMAFGAzgLkWEBbAIwFMAnAbQF0BKRAb0QF8BYAKFYgXSkQEMSTsZAIwAaRACYKiALyJhYya1ao+JKgG5EAei2IA7nBIBrdAEIgA)

**Comment by Zemnmez** [comment]
looks like this is issue https://github.com/microsoft/TypeScript/issues/25516

**Comment by danieldietrich** [comment]
Yes, it is. I would love to see more precise type inference.
